contracts/utils/RateLimited.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "../refs/CoreRef.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+/// @title abstract contract for putting a rate limit on how fast a contract can perform an action e.g. Minting
+/// @author Fei Protocol
+abstract contract RateLimited is CoreRef {
+
+    /// @notice maximum rate limit per second governance can set for this contract
+    uint256 public immutable MAX_RATE_LIMIT_PER_SECOND;
+
+    /// @notice the rate per second for this contract
+    uint256 public rateLimitPerSecond;
+
+    /// @notice the last time the buffer was used by the contract
+    uint256 public lastBufferUsedTime;
+
+    /// @notice the cap of the buffer that can be used at once
+    uint256 public bufferCap;
+
+    /// @notice a flag for whether to allow partial actions to complete if the buffer is less than amount
+    bool public doPartialAction;
+
+    uint256 private _bufferStored;
```

Comment:
doc

Reply:
Not conventionally needed on private vars but probably good to add here